middle_end/flambda2/simplify/inlining/inlining_report.ml

```diff
@@ -4,128 +4,660 @@
+        Format.fprintf ppf "[[file:%s::%s][in compilation unit %a]]" t file
+          Compilation_unit.print cu_uid
+      with Not_found ->
+        Format.fprintf ppf
+          "in compilation unit %a but no inlining reports were generated for \
+           this unit"
+          Compilation_unit.print cu_uid
+end
+
+module Inlining_tree = struct
+  module Key = struct
+    type scope_type =
+      | Module
+      | Class
+      | Unknown
+
+    let compare_scope_type a b =
+      match a, b with
+      | Module, Module | Class, Class | Unknown, Unknown -> 0
+      | Module, (Class | Unknown) | Class, Unknown -> 1
+      | (Class | Unknown), Module | Unknown, Class -> -1
+
+    type element =
+      | Fundecl of string
+      | Scope of scope_type * string
+      | Call of Inlining_history.Absolute.t
+
+    type t = Debuginfo.t * element
+
+    let compare_element a b =
+      match a, b with
+      | Fundecl s1, Fundecl s2 -> String.compare s1 s2
+      | Call s1, Call s2 -> Inlining_history.Absolute.compare s1 s2
+      | Scope (t1, s1), Scope (t2, s2) ->
+        let c = compare_scope_type t1 t2 in
+        if c <> 0 then c else String.compare s1 s2
+      | Fundecl _, (Call _ | Scope _) | Call _, Scope _ -> 1
+      | (Call _ | Scope _), Fundecl _ | Scope _, Call _ -> -1
+
+    let compare (dbg1, e1) (dbg2, e2) =
+      let c = Debuginfo.compare dbg1 dbg2 in
+      if c <> 0 then c else compare_element e1 e2
+  end
+
+  module Map = Map.Make (Key)
+
+  type item =
+    | Fundecl of fundecl
+    | Call of call
+    | Scope of t
+
+  and t = item Map.t
+
+  and fundecl =
+    { before_simplify : Decision_with_context.t option;
+      body : t;
+      after_simplify : Decision_with_context.t option;
+      uid : Uid.t
+    }
+
+  and call =
+    { decision_with_context : Decision_with_context.t option;
+      inlined : t;
+      uid : Uid.t
+    }
+
+  let empty = Map.empty
+
+  let insert_or_update_scope ~scope_type ~name ~cont t =
+    let key = Debuginfo.none, Key.Scope (scope_type, name) in
+    let m =
+      if not (Map.mem key t)
+      then Map.empty
+      else
+        match Map.find key t with
+        | Scope m -> m
+        | Call _ | Fundecl _ -> assert false
+    in
+    Map.add key (Scope (cont m)) t
+
+  let join_decision ~error:_ old next =
+    match old, next with
+    | Some _old, Some next ->
+      (* CR poechsel: This case should be invalid (hitting it would mean that we
+         took several decisions for the same exact path, which either indicates
+         that the inlining history is buggy or that we are simplifying the same
+         exact term several times in a row) but it seems that it can be
+         currently reached. I should investigate it a bit further.
+
+         Misc.fatal_errorf error *)
+      Some next
+    | None, next -> next
+    | old, None -> old
+
+  let insert_or_update_call ?decision_with_context ~dbg ~callee
+      ~(cont : item Map.t -> item Map.t) t =
```

Comment:
I'm not sure I'd call this a continuation - it's not tail-called, certainly not with the result of this function. I think this would actually be clearer with the argument labeled just `f` (see `Map.update` in Base). Also, is there a distinction intended between `t` and `item Map.t`?

Reply:
`item Map.t` is `t` so I removed occurrences to `item Map.t`